File: src/ddiff.c
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include "../include/ddiff.h"
prec_t * divided_difference(prec_t * x,prec_t *y, int num_ele) {
	prec_t * out=malloc(sizeof(prec_t)*num_ele);
	for (int i=0; i < num_ele; ++i)
		out[i]=y[i];
	for (int i=1; i < num_ele+1; ++i) {
		for (int j=num_ele-1; j > i; j--) {
			prec_t num=out[j]-out[j-1];
			prec_t den=x[j]-x[j-i];
			out[j]=num/den;
#if DEBUG_PRINTS
			printf("i=%d j=%d d=%e\n",i,j,out[j]);
#endif
		}
	}
	out[num_ele-1]=(out[num_ele-1]-out[num_ele-2])/(x[num_ele-1]-x[0]);
	return out;
}

int64_t fac(int x) {
	return (x==0?1:x*fac(x-1));
}

void scale_to_taylor(prec_t * terms, int num_ele) {
	for (int i=0; i < num_ele; ++i) {
		terms[i] = terms[i]/fac(i);
	}
}

void print_mac(float *c, int numele) {
	printf("[");
	for (int i=0; i < numele; ++i) {
		float coeff=c[i]/fac(i);
		printf("%e,",coeff);
	}
	printf("]\n");
}

int test_ddiff() {
	prec_t * x=malloc(sizeof(prec_t)*10);
	prec_t * y=malloc(sizeof(prec_t)*10);
	float bot=2e6;
	for (int i=0; i < 10; ++i) {
		x[i]=bot+(i*20);
		y[i]=x[i];
	}
	prec_t * dd=divided_difference(x, y, 10);
	printf("x\t\ty\t\tdd\n");
	for (int i=0; i < 10; ++i) {
		printf("%e\t%e\t%e\n",x[i],y[i],dd[i]);
	}
	return 0;
}

File: py/ddiff.py
from typing import Iterable
from math import atan, factorial

def ddiff(x:Iterable, y:Iterable,ne:int=-1):
	out=[i for i in y]
	ne=len(out)
	x=tuple(x)
	for i in range(1,ne+1):
		for j in range(ne-1,i-1,-1):
			# print(f"(y[{j}]-y[{j-1}])/(x[{j}]-x[{j-i}])",end=" = ")
			# print(f"({out[j]:1.4f}-{out[j-1]:1.4f})/({x[j]:1.1f}-{x[j-i]:1.1f})",end =" = ")
			num = out[j]-out[j-1]
			denom = x[j]-x[j-i]
			# print(f"{num:1.4f}/{denom:1.1f}={num/denom:1.4f}")
			out[j]=num/denom
	out[-1]=(out[-1]-out[-2])/(x[-1]-x[0])
	return out

def scale_to_taylor(terms:list[float]):
	return list(map(lambda i:terms[i]/factorial(i),range(len(terms))))

def L(f):
	a=1e-9
	b=2.8e-9
	c=800e-9
	f0=2e4
	res=(0.6366*a)*atan(-c*(f-f0))+b;
	return res
if __name__=="__main__":
	start=int(2e6)
	step=20
	ne=10
	xs=list(range(start,start+(ne*step),step))
	ys=tuple(map(L,xs))
	dd=ddiff(xs,ys)
	print("x\t\ty\t\tdd")
	for a,b,c in zip(xs,ys,dd):
		print(f"{a:e}\t{b:e}\t{c:e}")
	# print(ddiff(xs,ys))
